core/includes/updater.inc

```diff
@@ -175,19 +175,26 @@ abstract class Updater implements BackdropUpdaterInterface {
   }
 
   /**
-   * Return the project type from a Backdrop info file.
+   * Return the project type from a Backdrop info file or directory.
    *
    * @param string $directory
-   *   Directory to search for the info file.
+   *   Stream wrapper to directory to get the type from.
    *
    * @return string
-   *   The type of the project, either "module", "theme", or "layout".
+   *   The type of the project, either "core", "module", "theme", or "layout".
    *
    * @throws UpdaterException
    */
   public static function getProjectType($directory) {
-    if ($directory == BACKDROP_ROOT) {
-      return 'core';
+    if (basename($directory) == 'backdrop') {
+      $realpath = backdrop_realpath($directory);
+      if (is_dir($realpath . '/core')) {
+        return 'core';
+      }
+      else {
+        $message = 'Updating core from @directory was attempted, but either it doesn\'t contain a "core" directory, or it isn\'t accessible. You may need to update manually.';
+        throw new UpdaterException(t($message, array('@directory' => $directory)));
```

Comment:
This may be a technicality, but we usually would do `$message = t(...);` and then 
```suggestion
        throw new UpdaterException($message);
```

...point being that the actual message and the array that holds placeholders should stick together.

Reply:
Agreed, keeping them together makes sense.